optaplanner-examples/src/main/java/org/optaplanner/examples/examination/optional/score/ExaminationConstraintProvider.java

```diff
@@ -0,0 +1,369 @@
+                                  equal(Exam::getPeriod, Exam::getPeriod),
+                                  equal(Exam::getRoom, Exam::getRoom),
+                                  lessThan(Exam::getId, Exam::getId))
+                .join(Exam.class,
+                      equal(Exam::getPeriod),
+                      equal(Exam::getRoom),
+                      greaterThan(Exam::getId),
+                      filtering((leftExam, rightExam) -> rightExam.getTopicDuration() != leftExam.getTopicDuration()))
+                .ifNotExists(Exam.class,
+                             equal((leftExam, rightExam) -> rightExam.getPeriod(), Exam::getPeriod),
+                             equal((leftExam, rightExam) -> rightExam.getRoom(), Exam::getRoom),
+                             equal((leftExam, rightExam) -> rightExam.getTopicDuration(), Exam::getTopicDuration),
+                             greaterThan((leftExam, rightExam) -> rightExam.getId(), Exam::getId))
+                .penalize("Mixed durations", HardSoftScore.ONE_SOFT);
+    }
+
+    //    // Larger Exams towards the beginning of the examination session
+//    rule "frontLoad"
+//    when
+//    $exam : Exam(topicFrontLoadLarge == true, periodFrontLoadLast == true)
+//    then
+//        scoreHolder.penalize(kcontext);
+//    end
+    private Constraint frontLoad(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Exam.class)
+                .filter(exam -> exam.isTopicFrontLoadLarge() && exam.isPeriodFrontLoadLast())
+                .penalize("Front load", HardSoftScore.ONE_SOFT);
+    }
+
+    //    // Period Penalty
+//    rule "periodPenalty"
+//    when
+//    $period : Period(penalty != 0)
+//    $exam: Exam(period == $period)
+//    then
+//        scoreHolder.addSoftConstraintMatch(kcontext, - $period.getPenalty());
+//    end
+    private Constraint periodPenalty(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Period.class)
+                .filter(period -> period.getPenalty() != 0)
+                .ifExists(Exam.class, equal(Function.identity(), Exam::getPeriod))
+                .penalize("Period penalty", HardSoftScore.ONE_SOFT, Period::getPenalty);
+    }
+
+    //    // Room Penalty
+//    rule "roomPenalty"
+//    when
+//    $room : Room(penalty != 0)
+//    $exam: Exam(room == $room)
+//    then
+//        scoreHolder.addSoftConstraintMatch(kcontext, - $room.getPenalty());
+//    end
+    private Constraint roomPenalty(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Room.class)
+                .filter(room -> room.getPenalty() != 0)
+                .ifExists(Exam.class, equal(Function.identity(), Exam::getRoom))
+                .penalize("Room penalty", HardSoftScore.ONE_SOFT, Room::getPenalty);
+    }
+
+    private int getAbsoluteValue(Exam leftExam, Exam rightExam) {
+        return Math.abs(leftExam.getPeriodIndex() - rightExam.getPeriodIndex());
+    }
```

Comment:
Choose a better name for the method. Don't describe the code, describe the purpose.